File: win-prog/ProgWinVB/Lines Curves and Area Fills/PolyEllipse/PolyEllipse.vb
'--------------------------------------------
' PolyEllipse.vb (c) 2002 by Charles Petzold
'--------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class PolyEllipse
    Inherits PrintableForm

    Shared Shadows Sub Main()
        Application.Run(New PolyEllipse())
    End Sub

    Sub New()
        Text = "Ellipse with DrawLines"
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        Dim iNum As Integer = 2 * (cx + cy)
        Dim aptf(iNum) As PointF
        Dim i As Integer

        For i = 0 To iNum
            Dim rAng As Double = i * 2 * Math.PI / iNum
            aptf(i).X = CSng((cx - 1) / 2 * (1 + Math.Cos(rAng)))
            aptf(i).Y = CSng((cy - 1) / 2 * (1 + Math.Sin(rAng)))
        Next i

        grfx.DrawLines(New Pen(clr), aptf)
    End Sub
End Class

File: win-prog/ProgWinCS/Lines Curves and Area Fills/PolyEllipse/PolyEllipse.cs
//------------------------------------------
// PolyEllipse.cs © 2001 by Charles Petzold
//------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class PolyEllipse: PrintableForm
{
     public new static void Main()
     {
          Application.Run(new PolyEllipse());
     }
     public PolyEllipse()
     {
          Text = "Ellipse with DrawLines";
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          int      iNum = 2 * (cx + cy);
          PointF[] aptf = new PointF[iNum];

          for (int i = 0; i < iNum; i++)
          {
               double dAng = i * 2 * Math.PI / (iNum - 1);

               aptf[i].X = (cx - 1) / 2f * (1 + (float)Math.Cos(dAng));
               aptf[i].Y = (cy - 1) / 2f * (1 + (float)Math.Sin(dAng));
          }
          grfx.DrawLines(new Pen(clr), aptf);
     }
}
